src/track/track.cpp

```diff
@@ -1454,30 +1454,56 @@ CoverInfo Track::getCoverInfoWithLocation() const {
     return CoverInfo(m_record.getCoverInfo(), m_fileAccess.info().location());
 }
 
+namespace {
+
+bool safeToOverwriteMetadataOnExport(mixxx::TrackRecord::SourceSyncStatus sourceSyncStatus) {
+    return sourceSyncStatus == mixxx::TrackRecord::SourceSyncStatus::Synchronized ||
+            // After the migration the sync status of all existing tracks in the
+            // library is unknown. In this case we assume that exporting metadata
+            // is desired independent of the last modification time of the file.
+            sourceSyncStatus == mixxx::TrackRecord::SourceSyncStatus::Unknown ||
+            // Trying to write into an inaccessible file is safe even though it
+            // might fail subsequently. We need to continue here for avoiding
+            // misleading or wrong error messages about the synchronization status!
+            sourceSyncStatus == mixxx::TrackRecord::SourceSyncStatus::Undefined;
+}
+
+} // anonymous namespace
+
 ExportTrackMetadataResult Track::exportMetadata(
         const mixxx::MetadataSource& metadataSource,
         const SyncTrackMetadataParams& syncParams) {
     // Locking shouldn't be necessary here, because this function will
     // be called after all references to the object have been dropped.
     // But it doesn't hurt much, so let's play it safe ;)
     auto locked = lockMutex(&m_qMutex);
-    // TODO(XXX): Use sourceSynchronizedAt to decide if metadata
-    // should be (re-)imported before exporting it. The file might
-    // have been updated by external applications. Overwriting
-    // this modified metadata is probably not be intended if not
-    // explicitly requested by m_bMarkedForMetadataExport.
-    if (!m_bMarkedForMetadataExport &&
-            m_record.checkSourceSyncStatus(m_fileAccess.info()) !=
-                    mixxx::TrackRecord::SourceSyncStatus::Synchronized) {
-        // If the metadata has never been imported from file tags it
-        // must be exported explicitly once. This ensures that we don't
-        // overwrite existing file tags with completely different
-        // information.
-        kLogger.info()
-                << "Skip exporting of unsynchronized track metadata:"
-                << getLocation();
-        // abort
-        return ExportTrackMetadataResult::Skipped;
+    const auto sourceSyncStatus = m_record.checkSourceSyncStatus(m_fileAccess.info());
+    if (!safeToOverwriteMetadataOnExport(sourceSyncStatus)) {
```

Comment:
This is hard to understand because of double inversion. 
How about 
```suggestion
    if (isLibraryMetadataOutdated(sourceSyncStatus)) {
```

Is the "Void" case correctly handled? 
When does "Void" happen? If the file has no Metadata? 
Will it trigger an assertion?

Reply:
The Void case is handled as before, the behavior didn't change.

A test for this case could be added later by someone else. I don't have any interest to do this.

Reply:
Double inversion? No, that's not the case here.

The naming of this utility function has deliberately been chosen as is, because *outdated* is just one of the possible states were exporting metadata is considered to be safe. The name of the function simply reflects the intention of a complex if-condition.

Reply:
At this point we have to make a decision if exporting metadata should be permitted or not. The function hides the implementation details, i.e. checking for outdated metadata and more.

Reply:
-> Martin Fowler - Refactoring, introductory chapters.

Reply:
I have also thought about using a verbose switch statement inside this function and am missing Rust's exhaustive match statements here. Using a switch statement inside a utility function is a valid demand. But I refuse to drop the utility function and give up the readability improvements.